### DISCUSSION TITLE
Query `edge_logs` from SQL in the DB

### DISCUSSION THREAD

**Discussion by FelixZY** [supabase/supabase#8175] [Questions] [answered]
I'm building an API service where an app will connect to my supabase project using an API key. As part of this process, I want to restrict access if the API key has been used too much recently. I'm thinking the `edge_logs` is probably the best place to look but can't find a way to query them from within SQL.

Is there a way to query the edge/API logs using SQL from within the database, i.e. something like:

```sql
CREATE FUNCTION is_below_limit (key_id uuid) returns boolean AS $value$
BEGIN
  RETURN (
    SELECT COUNT(*) < 100 
    FROM `edge_logs` CROSS JOIN /* ... */
    WHERE id = key_id AND timestamp > /* ... */
  );
END;
$value$ LANGUAGE plpgsql;

**Comment by Ziinc** [comment, marked as answer]
Hi @FelixZY this feature is currently not supported, though it is definitely something that we are considering, as we definitely understand the use case.

For now, for your use case I would recommend counter-based restrictions instead of attempting to query the logs.

**Reply by FelixZY** [reply]
@Ziinc could you please elaborate on what you mean with "counter based restrictions"?

My current implementation is to increment a counter every time an API key is used. However, this is problematic since most of the API requests are GET requests, causing a read-only state.

I've gotten around this by making a new POST API call from within the database but this is definitely less than desirable as it causes a lot of overhead.

**Reply by Ziinc** [reply]
You could expose a function that triggers the read + update, then execute the function with `.rpc()` from the client side.

if your api requires an endpoint,you can perform the operations within an edge function as well. 

There is quite a lot of flexibility depending on your requirements.

**Reply by FelixZY** [reply]
Thank you for the ideas. However, forcing my api to consist solely of functions is a bit limiting.

Currently I'm using this implementation: https://gist.github.com/FelixZY/0aef530690458b381b8100afa19202c8

I have also created a feature request here: https://github.com/supabase/supabase/issues/7186

I would love to discuss the topic of API keys with you closer. This is a use case for many more than me but there are some platform limitations that make them cumbersome to work with right now.
